sim-lib/src/sim_node.rs

```diff
@@ -0,0 +1,1011 @@
+            return Err(ForwardingError::InsufficientBalance(
+                htlc.amount_msat,
+                self.local_balance_msat,
+            ));
+        }
+
+        if htlc.amount_msat < self.policy.min_htlc_size_msat {
+            return Err(ForwardingError::LessThanMinimum(
+                htlc.amount_msat,
+                self.policy.min_htlc_size_msat,
+            ));
+        }
+
+        if htlc.amount_msat > self.policy.max_htlc_size_msat {
+            return Err(ForwardingError::MoreThanMaximum(
+                htlc.amount_msat,
+                self.policy.max_htlc_size_msat,
+            ));
+        }
+
+        if self.in_flight.len() as u64 + 1 > self.policy.max_htlc_count {
+            return Err(ForwardingError::ExceedsInFlightCount(
+                self.in_flight.len() as u64,
+                self.policy.max_htlc_count,
+            ));
+        }
+
+        if self.in_flight_total() + htlc.amount_msat > self.policy.max_in_flight_msat {
+            return Err(ForwardingError::ExceedsInFlightTotal(
+                htlc.amount_msat,
+                self.in_flight_total(),
+                self.policy.max_in_flight_msat,
+            ));
+        }
+
+        if htlc.cltv_expiry > 500000000 {
+            return Err(ForwardingError::ExpiryInSeconds(htlc.cltv_expiry));
+        }
+
+        Ok(())
+    }
+
+    fn add_outgoing_htlc(&mut self, htlc: Htlc) -> Result<(), ForwardingError> {
+        self.check_policy(&htlc)?;
+
+        match self.in_flight.get(&htlc.hash) {
+            Some(_) => Err(ForwardingError::PaymentHashExists(htlc.hash)),
+            None => {
+                self.local_balance_msat -= htlc.amount_msat;
+                self.in_flight.insert(htlc.hash, htlc);
+                Ok(())
+            }
+        }
+    }
+
+    fn remove_outgoing_htlc(
+        &mut self,
+        hash: PaymentHash,
+        success: bool,
+    ) -> Result<Htlc, ForwardingError> {
+        match self.in_flight.remove(&hash) {
+            Some(v) => {
+                // If the HTLC failed, pending balance returns to local balance.
+                if !success {
+                    self.local_balance_msat += v.amount_msat
+                }
+
+                Ok(v)
+            }
+            None => Err(ForwardingError::PaymentHashNotFound(hash)),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SimulatedChannel {
+    capacity_msat: u64,
+    short_channel_id: u64,
+    node_1: ChannelState,
+    node_2: ChannelState,
+}
+
+impl SimulatedChannel {
+    pub fn new(
+        capacity_msat: u64,
+        short_channel_id: u64,
+        node_1: ChannelPolicy,
+        node_2: ChannelPolicy,
+    ) -> Self {
+        SimulatedChannel {
+            capacity_msat,
+            short_channel_id,
+            node_1: ChannelState::new(node_1, capacity_msat),
+            node_2: ChannelState::new(node_2, capacity_msat),
+        }
+    }
+}
+
+impl SimulatedChannel {
```

Comment:
Why have another implementation of `SimulatedChannel` if there isn't a huge difference between the concerns in both `impl`s?

Reply:
Mistake!